projects/cdk/directives/auto-focus/handlers/ios.handler.ts

```diff
@@ -37,42 +42,78 @@ export class TuiIosAutofocusHandler extends AbstractTuiAutofocusHandler {
 
-        const blurHandler = (): void => fakeInput.focus();
+        const blurHandler = (): void => fakeInput.focus({preventScroll: true});
         const focusHandler = (): void => {
-            setTimeout(() => {
-                this.element.focus();
-
-                /**
-                 * @note:
-                 * We can't remove the element immediately, because it breaks flow
-                 */
-                setTimeout(() => {
-                    fakeInput.removeEventListener('blur', blurHandler);
-                    fakeInput.removeEventListener('focus', focusHandler);
+            clearTimeout(this.fakeFocusTimeoutId);
+
+            this.fakeFocusTimeoutId = this.windowRef.setTimeout(() => {
+                clearTimeout(this.elementFocusTimeoutId);
+
+                fakeInput.removeEventListener('blur', blurHandler);
+                fakeInput.removeEventListener('focus', focusHandler);
+
+                this.elementFocusTimeoutId = this.windowRef.setTimeout(() => {
+                    this.element.focus({preventScroll: false});
                     fakeInput.remove();
-                });
+                }, duration);
             });
         };
 
-        /**
-         * @note: ping-pong eager strategy hack
-         * After creating an element and bringing it into DOM,
-         * the browser automatically focuses on the invisible element.
-         * And then, after focus is triggered, we try to focus on target element, and if we managed to refocus,
-         * then we try to focus again on an invisible element, so that the keyboard slowly appears.
-         * This ping pong allows the keyboard to not overlap the modal window.
-         */
         fakeInput.addEventListener('blur', blurHandler, {once: true});
         fakeInput.addEventListener('focus', focusHandler);
 
-        this.element.parentElement?.appendChild(fakeInput);
+        if (this.insideDialog()) {
+            this.windowRef.document.body.appendChild(fakeInput);
+        } else {
+            this.element.parentElement?.appendChild(fakeInput);
+        }
+
+        fakeInput.focus({preventScroll: true});
+    }
+
+    /**
+     * @note:
+     * emulate textfield position in layout with cursor
+     * before focus to real textfield element
+     */
+    private makeFakeInput(): HTMLInputElement {
+        const fakeInput: HTMLInputElement = this.renderer.createElement('input');
+
+        fakeInput.style.height = px(this.element.clientHeight);
+        fakeInput.style.width = px(this.element.clientWidth / 2);
+        fakeInput.style.position = 'fixed';
+        fakeInput.style.opacity = '0';
+        fakeInput.style.fontSize = px(16); // disable possible auto zoom
+        fakeInput.readOnly = true; // prevent keyboard for fake input
+
+        // @note: emulate position cursor before focus to real textfield element
+        fakeInput.style.top = px(this.element.getBoundingClientRect().top);
```

Comment:
You shouldn't request client rect twice, it causes a reflow. Also clientHeight and clientWidth, just get them from the same clientRect - request it once and store in a constant.